test/types/schema.test.ts

```diff
@@ -630,3 +630,75 @@ function gh11987() {
   expectError(userSchema.path<'foo'>('name'));
   expectType<SchemaTypeOptions<string>>(userSchema.path<'name'>('name').OptionsConstructor);
 }
+
+function gh12030() {
+  const Schema1 = new Schema({
+    users: [
+      {
+        username: { type: String }
+      }
+    ]
+  });
+
+  expectType<{
+    users: {
+      username?: string
+    }[];
+  }>({} as InferSchemaType<typeof Schema1>);
+
+  const Schema2 = new Schema({
+    createdAt: { type: Date, default: Date.now }
+  });
+
+  expectType<{ createdAt: Date }>({} as InferSchemaType<typeof Schema2>);
+
+  const Schema3 = new Schema({
+    users: [
+      new Schema({
+        username: { type: String },
+        credit: { type: Number, default: 0 }
+      })
+    ]
+  });
+
+  expectType<{
+    users: Types.DocumentArray<{
+      credit: number;
+      username?: string;
+    }>;
+  }>({} as InferSchemaType<typeof Schema3>);
+
+
+  const Schema4 = new Schema({
+    data: { type: { role: String }, default: {} }
+  });
+
+  expectType<{ data: { role?: string } }>({} as InferSchemaType<typeof Schema4>);
+
+  const Schema5 = new Schema({
+    data: { type: { role: Object }, default: {} }
+  });
+
+  expectType<{ data: { role?: any } }>({} as InferSchemaType<typeof Schema5>);
+
+  const Schema6 = new Schema({
+    track: {
+      backupCount: {
+        type: Number,
+        default: 0
+      },
+      count: {
+        type: Number,
+        default: 0
+      }
+    }
+  });
+
+  expectType<{
+    track?: {
```

Comment:
As per discussion in #12030, this should ideally be `track: { backupCount: number; count: number; }`. But I don't think that's significant enough to block merging and shipping this PR, because it's a meaningful improvement. Feel free to put in a separate PR.